compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs

```diff
@@ -0,0 +1,207 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+/// We can improve verification performance by conveying semantic information
+/// about function pointer calls down to the underlying solver. In particular,
+/// method calls through Rust dynamic trait objects use a vtable (virtual method
+/// table) that stores method pointers for the concrete object type rather than
+/// the shared abstract trait type.
+///
+/// This file build a map of virtual call sites to all of the possible trait
+/// implementations that match that method and trait. We then can write out
+/// this information as function pointer restrictions, improving verification
+/// performance compared to heuristics that consider a wider set of possible
+/// function pointer targets.
+///
+/// For the current CBMC implementation of function restrictions, see:
+///     http://cprover.diffblue.com/md__home_travis_build_diffblue_cbmc_doc_architectural_restrict-function-pointer.html
+use crate::GotocCtx;
+use cbmc::goto_program::{Expr, Location, Stmt, Symbol, Type};
+use cbmc::InternedString;
+use cbmc::NO_PRETTY_NAME;
+use rmc_restrictions::{CallSite, PossibleMethodEntry, TraitDefinedMethod, VtableCtxResults};
+use rustc_data_structures::stable_map::FxHashMap;
+use tracing::debug;
+
+/// This structure represents data about the vtable that we construct
+/// Per trait, per method, which functions could virtual call sites
+/// possibly refer to?
+pub struct VtableCtx {
+    // Option to actually enable restrictions
+    pub emit_vtable_restrictions: bool,
+
+    // Map: (normalized trait name, method index) -> possible implementations
+    possible_methods: FxHashMap<TraitDefinedMethod, Vec<InternedString>>,
+
+    // All sites where a virtual call takes place
+    call_sites: Vec<CallSite>,
+
+    // Internal tracing of index needed for call site wrappers
+    call_site_global_idx: usize,
+}
+
+/// Constructor
+impl VtableCtx {
+    pub fn new(emit_vtable_restrictions: bool) -> Self {
+        debug!("Restricting vtable function pointers? {:?}", emit_vtable_restrictions);
+        Self {
+            emit_vtable_restrictions,
+            possible_methods: FxHashMap::default(),
+            call_sites: Vec::new(),
+            call_site_global_idx: 0,
+        }
+    }
+}
+
+/// Interface for codegen to add possible methods
+impl VtableCtx {
+    /// Add a possible implementation for a virtual method call.
+    pub fn add_possible_method(
+        &mut self,
+        trait_name: InternedString,
+        method: usize,
+        imp: InternedString,
+    ) {
+        assert!(self.emit_vtable_restrictions);
+        let key = TraitDefinedMethod { trait_name, vtable_idx: method };
+
+        if let Some(possibilities) = self.possible_methods.get_mut(&key) {
+            possibilities.push(imp);
```

Comment:
Is there any reason we might want to keep this list ordered?

Reply:
Not currently.